File: ios/Sources/BackgroundLocationPlugin/GeofenceReporter.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

/// Posts region crossings straight to a server, from native code.
///
/// A crossing usually happens with the app dead. iOS relaunches it in the
/// background for a few seconds to deliver the region event, and the webview
/// may never boot in that window — so a crossing that only reaches JavaScript
/// is a crossing the server hears about whenever the user next opens the app,
/// if ever. For anything that has to act on the crossing (an automatic break,
/// an end-of-day punch-out), that is too late; the report has to leave from
/// here.
///
/// Every report is queued before it is sent, and the queue is persisted, so a
/// crossing in a basement with no signal still arrives once the phone is next
/// woken with a connection — by the next crossing, by the app opening, or by
/// the plugin loading. Sends are strictly in order: the server decides breaks
/// from the sequence of exits and entries, and delivering an entry before the
/// exit it answers would invert its meaning.
final class GeofenceReporter {
    private static let queueKey = "geofence.reportQueue"
    /// Bounded, oldest dropped first: a queue behind a server that never
    /// answers must not grow without limit.
    static let maxQueued = 100
    /// A report the server keeps rejecting for reasons other than auth is
    /// dropped after this many attempts rather than blocking everything behind
    /// it forever.
    static let maxAttempts = 20

    private let store: KeyValueStore
    private let lock = NSLock()
    private var isFlushing = false

    /// Called with a region id when the server refuses its credential (401 or
    /// 403): the token was revoked — typically the employee punched out — so
    /// the region has nothing left to report to and should stop being watched.
    var onCredentialRevoked: (String) -> Void = { _ in }

    init(store: KeyValueStore) {
        self.store = store
    }

    func enqueue(regionID: String, report: GeofenceReportSpec, transition: GeofenceTransition) {
        lock.lock()
        var queue = loadQueue()
        queue.append(QueuedGeofenceReport(
            regionId: regionID, url: report.url, authToken: report.authToken,
            transition: transition, attempts: 0))
        if queue.count > GeofenceReporter.maxQueued {
            queue.removeFirst(queue.count - GeofenceReporter.maxQueued)
        }
        saveQueue(queue)
        lock.unlock()
    }

    /// Sends queued reports in order until one fails for a reason worth
    /// retrying. Safe to call at any time and from any thread; a flush already
    /// in progress makes this a no-op.
    func flush() {
        lock.lock()
        if isFlushing || loadQueue().isEmpty {
            lock.unlock()
            return
        }
        isFlushing = true
        lock.unlock()

        let taskId = GeofenceReporter.beginBackgroundTask()
        sendNext {
            GeofenceReporter.endBackgroundTask(taskId)
        }
    }

    private func sendNext(done: @escaping () -> Void) {
        lock.lock()
        let queue = loadQueue()
        guard let next = queue.first else {
            isFlushing = false
            lock.unlock()
            done()
            return
        }
        lock.unlock()

        guard let url = URL(string: next.url) else {
            removeFirst(matching: next)
            sendNext(done: done)
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 20
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("ElsapiensBackgroundLocation/1.0", forHTTPHeaderField: "User-Agent")
        if let token = next.authToken, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try? JSONEncoder().encode(next.transition)

        URLSession.shared.dataTask(with: request) { [weak self] _, response, error in
            guard let self = self else { done(); return }
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0

            if error == nil, (200..<300).contains(status) {
                self.removeFirst(matching: next)
                self.sendNext(done: done)
                return
            }
            if status == 401 || status == 403 {
                // The credential is gone and will not come back: nothing queued
                // for this region can ever be delivered, and watching it on
                // would only wake the app for crossings no one will receive.
                self.removeAll(forRegion: next.regionId)
                self.onCredentialRevoked(next.regionId)
                self.sendNext(done: done)
                return
            }

            // No network, a timeout, or a server error: keep it, in place, and
            // stop here — later reports must not overtake it.
            self.recordFailedAttempt(for: next)
            self.lock.lock()
            self.isFlushing = false
            self.lock.unlock()
            done()
        }.resume()
    }

    // MARK: - Queue

    private func removeFirst(matching entry: QueuedGeofenceReport) {
        lock.lock()
        var queue = loadQueue()
        if let index = queue.firstIndex(where: { $0.matches(entry) }) {
            queue.remove(at: index)
        }
        saveQueue(queue)
        lock.unlock()
    }

    private func removeAll(forRegion regionID: String) {
        lock.lock()
        saveQueue(loadQueue().filter { $0.regionId != regionID })
        lock.unlock()
    }

    private func recordFailedAttempt(for entry: QueuedGeofenceReport) {
        lock.lock()
        var queue = loadQueue()
        if let index = queue.firstIndex(where: { $0.matches(entry) }) {
            queue[index].attempts += 1
            if queue[index].attempts >= GeofenceReporter.maxAttempts {
                queue.remove(at: index)
            }
        }
        saveQueue(queue)
        lock.unlock()
    }

    private func loadQueue() -> [QueuedGeofenceReport] {
        guard let raw = store.string(forKey: GeofenceReporter.queueKey),
              let data = raw.data(using: .utf8),
              let decoded = try? JSONDecoder().decode([QueuedGeofenceReport].self, from: data)
        else { return [] }
        return decoded
    }

    private func saveQueue(_ queue: [QueuedGeofenceReport]) {
        guard let data = try? JSONEncoder().encode(queue) else { return }
        store.set(String(data: data, encoding: .utf8), forKey: GeofenceReporter.queueKey)
    }

    // MARK: - Background time

    #if canImport(UIKit)
    /// iOS gives an app relaunched for a region event only a few seconds;
    /// asking for background time keeps it alive long enough for the send.
    private static func beginBackgroundTask() -> UIBackgroundTaskIdentifier {
        var identifier: UIBackgroundTaskIdentifier = .invalid
        identifier = UIApplication.shared.beginBackgroundTask(withName: "GeofenceReport") {
            UIApplication.shared.endBackgroundTask(identifier)
            identifier = .invalid
        }
        return identifier
    }

    private static func endBackgroundTask(_ identifier: UIBackgroundTaskIdentifier) {
        guard identifier != .invalid else { return }
        UIApplication.shared.endBackgroundTask(identifier)
    }
    #else
    private static func beginBackgroundTask() -> Int { 0 }
    private static func endBackgroundTask(_ identifier: Int) {}
    #endif
}

/// Where a region's crossings are posted. The token is the caller's, sent as a
/// Bearer credential; the server is expected to answer 401/403 once it is no
/// longer valid, which stops the region.
struct GeofenceReportSpec: Codable {
    let url: String
    let authToken: String?
}

struct QueuedGeofenceReport: Codable {
    let regionId: String
    let url: String
    let authToken: String?
    let transition: GeofenceTransition
    var attempts: Int

    func matches(_ other: QueuedGeofenceReport) -> Bool {
        regionId == other.regionId
            && transition.timestamp == other.transition.timestamp
            && transition.transition == other.transition.transition
    }
}

File: android/src/main/java/com/elsapiens/backgroundlocation/GeofenceReporter.java
package com.elsapiens.backgroundlocation;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.work.BackoffPolicy;
import androidx.work.Constraints;
import androidx.work.ExistingWorkPolicy;
import androidx.work.NetworkType;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkManager;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.TimeUnit;

/**
 * Posts region crossings straight to a server, from native code.
 *
 * The Android counterpart to iOS's {@code GeofenceReporter}. A crossing is
 * delivered to a process Android started solely for the broadcast; the webview
 * has usually not booted and may never boot before the process is killed, so a
 * crossing that only reaches JavaScript is one the server hears about whenever
 * the user next opens the app. Anything that must act on the crossing — an
 * automatic break, an end-of-day punch-out — needs it to leave from here.
 *
 * Every report is queued (persisted) before it is sent, and sending runs as a
 * WorkManager job with a network constraint: a crossing in a basement with no
 * signal is delivered once the phone reconnects, with the app never opened.
 * Sends are strictly in order, because the server decides breaks from the
 * sequence of exits and entries.
 */
public final class GeofenceReporter {

    private static final String TAG = "GeofenceReporter";
    static final String KEY_QUEUE = "geofence.reportQueue";
    static final int MAX_QUEUED = 100;
    /** A report rejected this many times for non-auth reasons is dropped. */
    static final int MAX_ATTEMPTS = 20;
    private static final String WORK_NAME = "geofence-report-flush";

    private static final Object LOCK = new Object();

    private GeofenceReporter() {}

    /** Queues a crossing and schedules delivery. */
    public static void enqueue(Context context, String regionId, JSONObject report, JSONObject transition) {
        String url = report.optString("url", "");
        if (url.isEmpty()) {
            return;
        }
        synchronized (LOCK) {
            KeyValueStore store = new SharedPrefsKeyValueStore(context);
            JSONArray queue = load(store);
            try {
                JSONObject entry = new JSONObject();
                entry.put("regionId", regionId);
                entry.put("url", url);
                entry.put("authToken", report.optString("authToken", ""));
                entry.put("transition", transition);
                entry.put("attempts", 0);
                queue.put(entry);
            } catch (JSONException e) {
                Log.w(TAG, "could not queue geofence report", e);
                return;
            }
            // Oldest dropped first: a queue behind a server that never answers
            // must not grow without limit.
            while (queue.length() > MAX_QUEUED) {
                queue.remove(0);
            }
            store.putString(KEY_QUEUE, queue.toString());
        }
        schedule(context);
    }

    /** Schedules a flush for as soon as there is a network. */
    public static void schedule(Context context) {
        OneTimeWorkRequest request = new OneTimeWorkRequest.Builder(FlushWorker.class)
                .setConstraints(new Constraints.Builder()
                        .setRequiredNetworkType(NetworkType.CONNECTED)
                        .build())
                .setBackoffCriteria(BackoffPolicy.EXPONENTIAL, 30, TimeUnit.SECONDS)
                .build();
        // APPEND_OR_REPLACE keeps one chain: a crossing arriving while a flush is
        // running queues behind it instead of racing it out of order.
        WorkManager.getInstance(context.getApplicationContext())
                .enqueueUniqueWork(WORK_NAME, ExistingWorkPolicy.APPEND_OR_REPLACE, request);
    }

    /**
     * Sends queued reports in order. Returns true when the queue is empty,
     * false when a send failed for a reason worth retrying (the rest stay
     * queued behind it).
     */
    static boolean flush(Context context) {
        KeyValueStore store = new SharedPrefsKeyValueStore(context);
        while (true) {
            JSONObject next;
            synchronized (LOCK) {
                JSONArray queue = load(store);
                if (queue.length() == 0) {
                    return true;
                }
                next = queue.optJSONObject(0);
                if (next == null) {
                    queue.remove(0);
                    store.putString(KEY_QUEUE, queue.toString());
                    continue;
                }
            }

            int status = send(next);
            String regionId = next.optString("regionId", "");
            if (status >= 200 && status < 300) {
                removeFirst(store);
                continue;
            }
            if (status == 401 || status == 403) {
                // The credential is gone and will not come back — typically the
                // employee punched out. Nothing queued for this region can ever
                // be delivered, and watching it on would only wake the app for
                // crossings no one will receive.
                removeAllForRegion(store, regionId);
                new GeofenceManager(context, store).remove(regionId);
                continue;
            }
            // No network, a timeout, or a server error: keep it in place and
            // stop, so later reports cannot overtake it.
            recordFailedAttempt(store);
            return false;
        }
    }

    /** Returns the HTTP status, or 0 when the request never completed. */
    private static int send(JSONObject entry) {
        HttpURLConnection connection = null;
        try {
            connection = (HttpURLConnection) new URL(entry.getString("url")).openConnection();
            connection.setRequestMethod("POST");
            connection.setConnectTimeout(20_000);
            connection.setReadTimeout(20_000);
            connection.setDoOutput(true);
            connection.setRequestProperty("Content-Type", "application/json");
            connection.setRequestProperty("User-Agent", "ElsapiensBackgroundLocation/1.0");
            String token = entry.optString("authToken", "");
            if (!token.isEmpty()) {
                connection.setRequestProperty("Authorization", "Bearer " + token);
            }
            byte[] body = entry.getJSONObject("transition").toString().getBytes(StandardCharsets.UTF_8);
            try (OutputStream out = connection.getOutputStream()) {
                out.write(body);
            }
            return connection.getResponseCode();
        } catch (Exception e) {
            Log.w(TAG, "geofence report send failed: " + e.getMessage());
            return 0;
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }
    }

    // ── Queue ────────────────────────────────────────────────────────────────

    private static JSONArray load(KeyValueStore store) {
        String raw = store.getString(KEY_QUEUE, null);
        if (raw == null || raw.isEmpty()) {
            return new JSONArray();
        }
        try {
            return new JSONArray(raw);
        } catch (JSONException e) {
            return new JSONArray();
        }
    }

    private static void removeFirst(KeyValueStore store) {
        synchronized (LOCK) {
            JSONArray queue = load(store);
            if (queue.length() > 0) {
                queue.remove(0);
            }
            store.putString(KEY_QUEUE, queue.toString());
        }
    }

    private static void removeAllForRegion(KeyValueStore store, String regionId) {
        synchronized (LOCK) {
            JSONArray queue = load(store);
            JSONArray kept = new JSONArray();
            for (int i = 0; i < queue.length(); i++) {
                JSONObject entry = queue.optJSONObject(i);
                if (entry != null && !regionId.equals(entry.optString("regionId"))) {
                    kept.put(entry);
                }
            }
            store.putString(KEY_QUEUE, kept.toString());
        }
    }

    private static void recordFailedAttempt(KeyValueStore store) {
        synchronized (LOCK) {
            JSONArray queue = load(store);
            JSONObject first = queue.optJSONObject(0);
            if (first == null) {
                return;
            }
            int attempts = first.optInt("attempts", 0) + 1;
            if (attempts >= MAX_ATTEMPTS) {
                queue.remove(0);
            } else {
                try {
                    first.put("attempts", attempts);
                } catch (JSONException ignored) {
                    // attempts is a plain int; put cannot fail here
                }
            }
            store.putString(KEY_QUEUE, queue.toString());
        }
    }

    /** Runs the flush off the main thread, retried with backoff by WorkManager. */
    public static class FlushWorker extends Worker {
        public FlushWorker(@NonNull Context context, @NonNull WorkerParameters params) {
            super(context, params);
        }

        @NonNull
        @Override
        public Result doWork() {
            return flush(getApplicationContext()) ? Result.success() : Result.retry();
        }
    }
}
